BedrockConflictMetrics.cpp

```diff
@@ -0,0 +1,103 @@
+#include "BedrockConflictMetrics.h"
+
+// Initialize non-const static variables.
+recursive_mutex BedrockConflictMetrics::_mutex;
+map<string, BedrockConflictMetrics> BedrockConflictMetrics::_conflictInfoMap;
+
+BedrockConflictMetrics::BedrockConflictMetrics(const string& commandName) :
+_commandName(commandName)
+{ }
+
+void BedrockConflictMetrics::success() {
+    ++_totalSuccessCount;
+    _results.reset(_resultsPtr);
+    ++_resultsPtr;
+    _resultsPtr %= COMMAND_COUNT;
+}
+
+void BedrockConflictMetrics::conflict() {
+    ++_totalConflictCount;
+    _results.set(_resultsPtr);
+    ++_resultsPtr;
+    _resultsPtr %= COMMAND_COUNT;
+}
+
+int BedrockConflictMetrics::recentSuccessCount() {
+    return COMMAND_COUNT - _results.count();
+}
+
+int BedrockConflictMetrics::recentConflictCount() {
+    return _results.count();
+}
+
+void BedrockConflictMetrics::recordConflict(const string& commandName) {
+    SAUTOLOCK(_mutex);
+
+    // Look up, and create if necessary, the info object for this command.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // We need to insert this command.
+        auto itPair = _conflictInfoMap.emplace(make_pair(commandName, commandName));
+        it = itPair.first;
+    }
+    it->second.conflict();
+    SINFO("Multi-write conflict recorded for " << commandName);
+}
+
+void BedrockConflictMetrics::recordSuccess(const string& commandName) {
+    SAUTOLOCK(_mutex);
+
+    // Look up, and create if necessary, the info object for this command.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // We need to insert this command.
+        auto itPair = _conflictInfoMap.emplace(make_pair(commandName, commandName));
+        it = itPair.first;
+    }
+    it->second.success();
+    SINFO("Multi-write success recorded for " << commandName);
+}
+
+bool BedrockConflictMetrics::multiWriteOK(const string& commandName) {
+    SAUTOLOCK(_mutex);
+    bool result;
+
+    // Look up the command in the list of commands.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // If we don't find it, we assume it's OK to multi-write.
+        SINFO("Multi-write command '" << commandName << "' not tracked in BedrockConflictMetrics. Assuming OK.");
+        result = true;
+    } else {
+        // Otherwise, we check to see if it's recent conflict count is too high for multi-write.
+        auto& metric = it->second;
+        int conflicts = metric.recentConflictCount();
+        uint64_t totalAttempts = metric._totalConflictCount + metric._totalSuccessCount;
```

Comment:
Why bother with `totalSuccessCount` versus just initialize the bit vector to all "success"?  This would require 10 conflicts before it gets blocked, but I think that's ok.  Just seems needlessly complex, or maybe I'm missing what problem it tries to solve.